<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a summary of match filtering — create_summary_res • posDemux</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a summary of match filtering — create_summary_res"><meta name="description" content="create_summary_res() is a helper function in order
to create a summary of the demultiplexing
and following match filtering. It is not designed to be invoked directly, but
its results will be returned automatically from
filter_demultiplex_res(). This returned object has it own method
for printing the result in a user-friendly manner."><meta property="og:description" content="create_summary_res() is a helper function in order
to create a summary of the demultiplexing
and following match filtering. It is not designed to be invoked directly, but
its results will be returned automatically from
filter_demultiplex_res(). This returned object has it own method
for printing the result in a user-friendly manner."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">posDemux</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/demultiplexing.html">Introduction to combinatorial demultiplexing</a></li>
    <li><a class="dropdown-item" href="../articles/streaming.html">Streaming data into the demultiplexer</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/yaccos/posDemux/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create a summary of match filtering</h1>
      <small class="dont-index">Source: <a href="https://github.com/yaccos/posDemux/blob/main/R/match_summary.R" class="external-link"><code>R/match_summary.R</code></a></small>
      <div class="d-none name"><code>create_summary_res.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>create_summary_res()</code> is a helper function in order
to create a summary of the demultiplexing
and following match filtering. It is not designed to be invoked directly, but
its results will be returned automatically from
<code><a href="filter_demultiplex_res.html">filter_demultiplex_res()</a></code>. This returned object has it own method
for printing the result in a user-friendly manner.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_summary_res</span><span class="op">(</span></span>
<span>  <span class="va">retained</span>,</span>
<span>  <span class="va">barcodes</span>,</span>
<span>  <span class="va">assigned_barcodes</span>,</span>
<span>  <span class="va">allowed_mismatches</span>,</span>
<span>  <span class="va">mismatches</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'demultiplex_filter_summary'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-retained">retained<a class="anchor" aria-label="anchor" href="#arg-retained"></a></dt>
<dd><p>Logical vector with the same length as
the number of reads in the input to the demultiplexer.
<code>TRUE</code> if the corresponding read
is retained. Corresponds to the field <code>retained</code> of the output of
<code><a href="filter_demultiplex_res.html">filter_demultiplex_res()</a></code>.</p></dd>


<dt id="arg-barcodes">barcodes<a class="anchor" aria-label="anchor" href="#arg-barcodes"></a></dt>
<dd><p>A list of
<code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">XStringSet</a></code> objects, the
barcodes which were used for demultiplexing.</p></dd>


<dt id="arg-assigned-barcodes">assigned_barcodes<a class="anchor" aria-label="anchor" href="#arg-assigned-barcodes"></a></dt>
<dd><p>Character matrix of the assigned barcodes only
including the onces within the mismatch threshold.
Corresponds to of the field <code>demultiplex_res$assigned_barcodes</code>
of <code><a href="filter_demultiplex_res.html">filter_demultiplex_res()</a></code>.</p></dd>


<dt id="arg-allowed-mismatches">allowed_mismatches<a class="anchor" aria-label="anchor" href="#arg-allowed-mismatches"></a></dt>
<dd><p>Integer vector of length one or the same length
as the number of barcode segments; the threshold Hamming distance. All reads
having a number of mismatches above this number in any of the barcodes will
be filtered away.</p></dd>


<dt id="arg-mismatches">mismatches<a class="anchor" aria-label="anchor" href="#arg-mismatches"></a></dt>
<dd><p>Integer matrix of the number of
mismatches of each assigned barcode.
Corresponds to the field <code>mismatches</code> of
<code><a href="combinatorial_demultiplex.html">combinatorial_demultiplex()</a></code>.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object of class <code>demultiplex_filter_summary</code> from
<code>create_summary_res()</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Ignored</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>create_summary_res()</code> returns a list of
S3 class <code>demultiplex_filter_summary</code>
providing diagnostics for the filtering process. It contains the
the following fields:</p><ul><li><p><code>n_reads</code>: The total number of reads
in the dataset before filtering.</p></li>
<li><p><code>n_removed</code>: The number of reads removed
because demultiplexing failed.</p></li>
<li><p><code>n_barcode_sets</code>: The number of barcode sets.</p></li>
<li><p><code>n_barcode_combinations</code>: The possible number
of barcode combinations.</p></li>
<li><p><code>n_unique_barcodes</code>: The number of observed
unique barcode combinations
(i.e. features which may be cells) detected after filtering mismatches.</p></li>
<li><p><code>n_estimated_features</code>: The estimated number of features having a
detected combination of barcodes.
This number will always be greater or equal than <code>n_unique_barcodes</code> due
to barcode collisions.</p></li>
<li><p><code>observed_collision_lambda</code>: The ratio of observed barcode
combinations divided by the total number of possible barcode combinations.</p></li>
<li><p><code>corrected_collision_lambda</code>: The ratio of
estimated number of features
to the total number of possible barcode combinations.</p></li>
<li><p><code>expected_collisions</code>: The statistically expected number
of barcode collisions or more precicely the expected number of
observed barcodes which correspond to two or more features.</p></li>
<li><p><code>barcode_summary</code>: A list containing a summary
for each barcode set.
Each element contains the following:</p><ul><li><p><code>width</code>: The width (number of nucleotides) of the barcode set.</p></li>
<li><p><code>n_barcodes</code>: Number of query barcodes.</p></li>
<li><p><code>n_allowed_mismatches</code>: Number of allowed mismatches
for the barcode set.</p></li>
<li><p><code>n_removed</code>: Number of reads having too many mismatches
for this barcode set.</p></li>
<li><p><code>mismatch_frame</code>: A <code>data.frame</code> with the two columns,
<code>n_mismatches</code> and <code>frequency</code> showing the number of reads for each
of the allowed number of mismatches for the given barcode set.</p></li>
</ul></li>
</ul><p>The <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> method returns its output invisibly.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Following a uniform distribution of barcodes, the expected number
of barcode collisions
(observed barcodes combinations being composed of two or more features)
is given by
$$N\left(1-e^{-\lambda}-\lambda e^{-\lambda}\right),$$
where \(N\) is the number of possible barcode combinations
and \(\lambda\) is in this summary referred to as the collision lambda:
$$\lambda=\frac{n}{N},$$ where \(n\) is the number of features.
However, \(n\) is unknown as we cannot know how many features
there were originally due to potential collisions.
Utilizing the fact that the expected observed number of barcodes is given by
$$N\left(1-e^{-\lambda}\right),$$
we can correct the estimate for \(\lambda\) from the known value
of the observed barcode combinations, and thus estimate the number
of features and barcode collisions.</p>
<p>While each unique feature can be conceptually
thought of as single cell with its transcripts,
realistic datasets have many features with relatively
small numbers of reads which
are artifacts and unlikely to correspond to true cells.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">input_fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"extdata"</span>, <span class="st">"PETRI-seq_forward_reads.fq.gz"</span>, package <span class="op">=</span> <span class="st">"posDemux"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">input_fastq</span>, format <span class="op">=</span> <span class="st">"fastq"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">barcode_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"extdata/PETRI-seq_barcodes"</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>bc1 <span class="op">=</span> <span class="st">"bc1.fa"</span>, bc2 <span class="op">=</span> <span class="st">"bc2.fa"</span>, bc3 <span class="op">=</span> <span class="st">"bc3.fa"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"posDemux"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">barcode_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"bc"</span>, <span class="fl">1L</span><span class="op">:</span><span class="fl">3L</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">barcode_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">barcode_files</span>, <span class="va">readDNAStringSet</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="va">barcode_index</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bc3"</span>, <span class="st">"bc2"</span>, <span class="st">"bc1"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">sequence_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>UMI <span class="op">=</span> <span class="st">"P"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">segment_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7L</span>, <span class="fl">7L</span>, <span class="fl">15L</span>, <span class="fl">7L</span>, <span class="fl">14L</span>, <span class="fl">7L</span>, <span class="cn">NA_integer_</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">demultiplex_res</span> <span class="op">&lt;-</span> <span class="fu">posDemux</span><span class="fu">::</span><span class="fu"><a href="combinatorial_demultiplex.html">combinatorial_demultiplex</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">reads</span>, barcodes <span class="op">=</span> <span class="va">barcodes</span>, segments <span class="op">=</span> <span class="va">sequence_annotation</span>,</span></span>
<span class="r-in"><span>    segment_lengths <span class="op">=</span> <span class="va">segment_lengths</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">filtered_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="filter_demultiplex_res.html">filter_demultiplex_res</a></span><span class="op">(</span><span class="va">demultiplex_res</span>, allowed_mismatches <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">freq_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_freq_table.html">create_freq_table</a></span><span class="op">(</span><span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">assigned_barcodes</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">filtered_res</span><span class="op">$</span><span class="va">summary_res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of reads: 56895 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads failing to demultiplex: 4989 (8.77%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Observed number of unique barcode combinations:978 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of possible barcode combinations: 884736 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Estimated number of features: 978.5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Observed feature to barcode ratio: 0.001105 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Corrected feature to barcode ratio: 0.001106 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Estimated number of observed barcode combinations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> corresponding to more than one feature: 0.5 (0.06%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of barcode sets: 3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Barcode set: bc3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Barcode width: 7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of possible barcodes: 96 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of allowed mismatches: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads with 0 mismatches: 51923 (91.26%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads with 1 mismatches: 671 (1.18%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads above mismatch threshold: 4301 (7.56%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Barcode set: bc2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Barcode width: 7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of possible barcodes: 96 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of allowed mismatches: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads with 0 mismatches: 51918 (91.25%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads with 1 mismatches: 567 (1%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads above mismatch threshold: 4410 (7.75%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Barcode set: bc1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Barcode width: 7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of possible barcodes: 96 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of allowed mismatches: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads with 0 mismatches: 51930 (91.27%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads with 1 mismatches: 642 (1.13%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of reads above mismatch threshold: 4323 (7.6%) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># This also works, but is usually not necessary to call directly</span></span></span>
<span class="r-in"><span><span class="va">alternative_summary_res</span> <span class="op">&lt;-</span> <span class="fu">create_summary_res</span><span class="op">(</span></span></span>
<span class="r-in"><span>    retained <span class="op">=</span> <span class="va">filtered_res</span><span class="op">$</span><span class="va">retained</span>, barcodes <span class="op">=</span> <span class="va">barcodes</span>,</span></span>
<span class="r-in"><span>    assigned_barcodes <span class="op">=</span> <span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">assigned_barcodes</span>,</span></span>
<span class="r-in"><span>    allowed_mismatches <span class="op">=</span> <span class="fl">1L</span>, mismatches <span class="op">=</span> <span class="va">demultiplex_res</span><span class="op">$</span><span class="va">mismatches</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jakob Peder Pettersen, Centre for new antibacterial strategies (CANS).</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

