<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to combinatorial demultiplexing • posDemux</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to combinatorial demultiplexing">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">posDemux</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/demultiplexing.html">Introduction to combinatorial demultiplexing</a></li>
    <li><a class="dropdown-item" href="../articles/streaming.html">Streaming data into the demultiplexer</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/yaccos/posDemux/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to combinatorial demultiplexing</h1>
                        <h4 data-toc-skip class="author">Jakob Peder
Pettersen</h4>
            <address class="author_afil">
      UiT, the Arctic University of Norway, Department of Computer
Science<br><a class="author_email" href="mailto:#"></a><a href="mailto:jakobpeder.pettersen@gmail.com" class="email">jakobpeder.pettersen@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">15 December 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/yaccos/posDemux/blob/main/vignettes/demultiplexing.Rmd" class="external-link"><code>vignettes/demultiplexing.Rmd</code></a></small>
      <div class="d-none name"><code>demultiplexing.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="basics">Basics<a class="anchor" aria-label="anchor" href="#basics"></a>
</h2>
<div class="section level3">
<h3 id="when-and-why-to-use-this-package">When and why to use this package<a class="anchor" aria-label="anchor" href="#when-and-why-to-use-this-package"></a>
</h3>
<p>This package is aimed at single-cell RNA-seq approaches such as
SPLiT-seq
<a name="cite-Rosenberg2018"></a><a name="cite-Kuijpers2024"></a>[<a href="https://www.science.org/doi/10.1126/science.aam8999" class="external-link">1</a>; <a href="https://doi.org/10.1186/s12864-024-10285-3" class="external-link">2</a>], BacDrop
<a name="cite-Ma2023"></a>[<a href="https://www.sciencedirect.com/science/article/pii/S0092867423000028" class="external-link">3</a>],
and PETRI-seq
<a name="cite-Blattman2020"></a><a name="cite-Blattman2024"></a>[<a href="https://www.nature.com/articles/s41564-020-0729-6" class="external-link">4</a>; <a href="https://www.nature.com/articles/s41586-024-08124-2" class="external-link">5</a>]. Here,
each cell to be sequenced is defined by its own unique combination of
multiple DNA barcodes. We do assume that the individual barcodes come
from a predefined whitelist, but that the combination of barcodes is
random and that every possible combination of barcodes is valid. In
order to identify from which cell each read is coming from, we must
extract the barcodes, compare them with the reference barcodes, and find
the best match.</p>
</div>
<div class="section level3">
<h3 id="required-knowledge">Required knowledge<a class="anchor" aria-label="anchor" href="#required-knowledge"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.23/posDemux" class="external-link">posDemux</a></em>
uses the <em><a href="https://bioconductor.org/packages/3.23/Biostrings" class="external-link">Biostrings</a></em>
package for handling sequencing data and hence a basic understanding of
<code>Biostrings</code> is necessary to use this package. Also important
to note is the fact that while <code>posDemux</code> does provide
utilities for demultiplexing scRNA-seq data, it is designed to be part
of a greater workflow. In particular, the package does not include
essential components of a scRNA-seq workflow like quality control,
alignment, feature counting and UMI deduplication, neither does it
provide any utilities for tertiary analysis once the gene count matrix
is constructed. Hence, the user should be familiar with the other
components of such workflows [<a href="https://doi.org/10.1186/s12864-024-10285-3" class="external-link">2</a>] and know how
the functionality of this package plays a part therein.</p>
<p>If you are asking yourself the question “Where do I start using
Bioconductor?” you might be interested in <a href="https://lcolladotor.github.io/2014/10/16/where-do-i-start-using-bioconductor/" class="external-link">this
blog post</a>.</p>
</div>
<div class="section level3">
<h3 id="limitations">Limitations<a class="anchor" aria-label="anchor" href="#limitations"></a>
</h3>
<p>Since this is a positional demultiplexer, the position of all
segments and barcodes must be known in advance and be the same for all
reads. Hence, the package is not suited in cases where segments occur at
variable positions among the reads. Furthermore, in the cases where the
segments are at fixed positions among the reads, but the segment lengths
are unknown, it may require some trial and error to obtain the correct
segmentation.</p>
<p>The demultiplexer is written with Illumina sequencing in mind. It
assumes that substitution sequencing errors may occur, but has no
safeguard against indel sequencing errors. Hence, indel errors will
result in erroneous segmentation and failure to assign the correct
barcode. For sequencing platforms where indel errors are common, such as
Nanopore sequencing, this package may therefore not be suitable.</p>
<p>This package does not by itself support combining barcodes on both
reverse and forward reads, but this limitation can be circumvented by
either artificially merging reads or manipulating the result structures
from the demultiplexer.</p>
</div>
<div class="section level3">
<h3 id="ways-to-interact-with-this-package">Ways to interact with this package<a class="anchor" aria-label="anchor" href="#ways-to-interact-with-this-package"></a>
</h3>
<p>This vignette describes the basic usage of the package where the
FASTQ files are read into <code>R</code> before the demultiplexing
begins. For large datasets or memory-limited systems, reading all
sequences into memory at once is too taxing. For this reason,
<code>posDemux</code> has a streaming API where only parts of the input
file is read at a time. For brevity, this functionality is explained in
its own vignette (<code><a href="../articles/streaming.html">vignette("streaming")</a></code>). We believe that
most end-users won’t interact with this package directly, but rather in
the form of a custom-made bioinformatic pipeline. Also, we think that
most developers of bioinformatics pipelines would prefer the streaming
API. Still, we think those developing with this package should have an
understanding of its low-level API as to understand what it does under
the hood.</p>
</div>
<div class="section level3">
<h3 id="install-posdemux">Install <code>posDemux</code><a class="anchor" aria-label="anchor" href="#install-posdemux"></a>
</h3>
<p>This package is hosted on <a href="http://bioconductor.org" class="external-link">Bioconductor</a> repository for
<code>R</code> packages. In order to install it from a fresh
<code>R</code> install, run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"posDemux"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Check that you have a valid Bioconductor installation</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/valid.html" class="external-link">valid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>If you want to obtain the development version of the package, it is
available on <a href="https://github.com/yaccos/posDemux" class="external-link">GitHub</a>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"yaccos/posDemux"</span><span class="op">)</span></span></code></pre></div>
<p>If you compile the package from source, we do recommend you use the
compiler flag <code>-ftree-vectorize</code> as this provides
considerable speedup to the demultiplexer.</p>
</div>
<div class="section level3">
<h3 id="asking-for-help">Asking for help<a class="anchor" aria-label="anchor" href="#asking-for-help"></a>
</h3>
<p>As package developers, we try to explain clearly how to use our
packages and in which order to use the functions. But <code>R</code> and
<code>Bioconductor</code> have a steep learning curve so it is critical
to learn where to ask for help. The blog post quoted above mentions some
but we would like to highlight the <a href="https://support.bioconductor.org/" class="external-link">Bioconductor support site</a>
as the main resource for getting help: Remember to use the
<code>posDemux</code> tag and check <a href="https://support.bioconductor.org/tag/posDemux/" class="external-link">the older
posts</a>. Other alternatives are available such as creating GitHub
issues and tweeting. However, please note that if you want to receive
help you should adhere to the <a href="http://www.bioconductor.org/help/support/posting-guide/" class="external-link">posting
guidelines</a>. It is particularly critical that you provide a small
reproducible example and your session information so package developers
can track down the source of the error.</p>
</div>
<div class="section level3">
<h3 id="citing-posdemux">Citing <code>posDemux</code><a class="anchor" aria-label="anchor" href="#citing-posdemux"></a>
</h3>
<p>We hope that <em><a href="https://bioconductor.org/packages/3.23/posDemux" class="external-link">posDemux</a></em>
will be useful for your research. Please use the following information
to cite the package and the overall approach. Thank you!</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Citation info</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"posDemux"</span><span class="op">)</span></span>
<span><span class="co">#&gt; To cite package 'posDemux' in publications use:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   Pettersen J (2025). _posDemux: Positional combinatorial sequence</span></span>
<span><span class="co">#&gt;   demultiplexer_. R package version 0.99.8,</span></span>
<span><span class="co">#&gt;   &lt;https://github.com/yaccos/posDemux&gt;.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   @Manual{,</span></span>
<span><span class="co">#&gt;     title = {posDemux: Positional combinatorial sequence demultiplexer},</span></span>
<span><span class="co">#&gt;     author = {Jakob Peder Pettersen},</span></span>
<span><span class="co">#&gt;     year = {2025},</span></span>
<span><span class="co">#&gt;     note = {R package version 0.99.8},</span></span>
<span><span class="co">#&gt;     url = {https://github.com/yaccos/posDemux},</span></span>
<span><span class="co">#&gt;   }</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="an-example-with-petri-seq">An example with PETRI-seq<a class="anchor" aria-label="anchor" href="#an-example-with-petri-seq"></a>
</h2>
<div class="section level3">
<h3 id="quick-overview-of-the-method">Quick overview of the method<a class="anchor" aria-label="anchor" href="#quick-overview-of-the-method"></a>
</h3>
<p>The PETRI-seq method is a method for single-cell RNA-sequencing of
bacteria [<a href="https://www.nature.com/articles/s41564-020-0729-6" class="external-link">4</a>; <a href="https://www.nature.com/articles/s41586-024-08124-2" class="external-link">5</a>]. It
uses paired-end Illumina sequencing where the forward read contains the
cell barcodes and the transcript UMI, whereas the reverse read contain
the cDNA to be aligned to the genome. Hence, <code>posDemux</code> will
only be applied to the forward reads. We start by loading the
<code>posDemux</code> itself and some helper packages:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yaccos/posDemux" class="external-link">posDemux</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sequence-annotation">Sequence annotation<a class="anchor" aria-label="anchor" href="#sequence-annotation"></a>
</h3>
<p>For PETRI-seq, the forward read consists of the following segments in
the following order (from 5’ to 3’):</p>
<ul>
<li>UMI: 7 nucletides</li>
<li>Barcode 3(<code>bc3</code>) : 7 nucleotides</li>
<li>Linker: 15 nucletides</li>
<li>Barcode 2(<code>bc2</code>): 7 nucleotides</li>
<li>Linker: 14 nucletides</li>
<li>Barcode 1(<code>bc1</code>): 7 nucleotides</li>
<li>The rest of the read is ignored</li>
</ul>
<p>Hence, we consider the first segment, the UMI, as a payload (denoted
as <code>P</code> in the sequence annotation) to be kept. The three
barcodes (denoted as <code>B</code>) are used for demultiplexing.
Finally, the linkers and the read past Barcode 1 (denoted as
<code>A</code>) are ignored.</p>
<p>We specify it with as:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequence_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>UMI <span class="op">=</span> <span class="st">"P"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span><span class="op">)</span></span>
<span><span class="va">segment_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7L</span>, <span class="fl">7L</span>, <span class="fl">15L</span>, <span class="fl">7L</span>, <span class="fl">14L</span>, <span class="fl">7L</span>, <span class="cn">NA_integer_</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h3>
<p>For the demultiplexing, we need the listing of the barcodes, and of
course the reads to demultiplex. For this package, the PETRI-seq
barcodes and synthetically generated reads are provided. We load the
barcodes:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barcode_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/PETRI-seq_barcodes"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        bc1 <span class="op">=</span> <span class="st">"bc1.fa"</span>,</span>
<span>        bc2 <span class="op">=</span> <span class="st">"bc2.fa"</span>,</span>
<span>        bc3 <span class="op">=</span> <span class="st">"bc3.fa"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"posDemux"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">barcode_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"bc"</span>, <span class="fl">1L</span><span class="op">:</span><span class="fl">3L</span><span class="op">)</span></span>
<span><span class="va">barcode_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">barcode_files</span>, <span class="va">readDNAStringSet</span><span class="op">)</span></span></code></pre></div>
<p>… and the FASTQ file containing the forward reads:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input_fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    <span class="st">"PETRI-seq_forward_reads.fq.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"posDemux"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">input_fastq</span>, format <span class="op">=</span> <span class="st">"fastq"</span><span class="op">)</span></span></code></pre></div>
<p><code>posDemux</code> accepts both <code>DNAStringSet</code> and
<code>QualityScaledDNAStringSet</code> as input where the latter contain
the quality scores of the reads. However, the quality score is ignored
by the demultiplexer and the only reason to pass a
<code>QualityScaledDNAStringSet</code> is when it is desirable to retain
the quality scores in the payload.</p>
</div>
<div class="section level3">
<h3 id="running-the-demultiplexer">Running the demultiplexer<a class="anchor" aria-label="anchor" href="#running-the-demultiplexer"></a>
</h3>
<p>Before we run the demultiplexer, we must ensure that the barcodes
index is arranged in the correct order:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="va">barcode_index</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bc3"</span>, <span class="st">"bc2"</span>, <span class="st">"bc1"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>We are now ready to run the demultiplexer:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demultiplex_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combinatorial_demultiplex.html">combinatorial_demultiplex</a></span><span class="op">(</span></span>
<span>    <span class="va">reads</span>,</span>
<span>    barcodes <span class="op">=</span> <span class="va">barcodes</span>,</span>
<span>    segments <span class="op">=</span> <span class="va">sequence_annotation</span>,</span>
<span>    segment_lengths <span class="op">=</span> <span class="va">segment_lengths</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The main results from this demultiplexer are the table of assigned
barcodes:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">assigned_barcodes</span><span class="op">)</span></span>
<span><span class="co">#&gt;       bc3      bc2      bc1     </span></span>
<span><span class="co">#&gt; seq_1 "bc3_57" "bc2_51" "bc1_94"</span></span>
<span><span class="co">#&gt; seq_2 "bc3_72" "bc2_94" "bc1_95"</span></span>
<span><span class="co">#&gt; seq_3 "bc3_23" "bc2_37" "bc1_85"</span></span>
<span><span class="co">#&gt; seq_4 "bc3_85" "bc2_81" "bc1_17"</span></span>
<span><span class="co">#&gt; seq_5 "bc3_49" "bc2_64" "bc1_77"</span></span>
<span><span class="co">#&gt; seq_6 "bc3_3"  "bc2_45" "bc1_33"</span></span></code></pre></div>
<p>the table of mismatches to the barcode:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">mismatches</span><span class="op">)</span></span>
<span><span class="co">#&gt;       bc3 bc2 bc1</span></span>
<span><span class="co">#&gt; seq_1   0   0   0</span></span>
<span><span class="co">#&gt; seq_2   0   0   0</span></span>
<span><span class="co">#&gt; seq_3   2   2   2</span></span>
<span><span class="co">#&gt; seq_4   0   0   0</span></span>
<span><span class="co">#&gt; seq_5   0   0   0</span></span>
<span><span class="co">#&gt; seq_6   0   0   0</span></span></code></pre></div>
<p>as well as the extracted UMI sequences:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">payload</span><span class="op">$</span><span class="va">UMI</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 56895:</span></span>
<span><span class="co">#&gt;         width seq                                           names               </span></span>
<span><span class="co">#&gt;     [1]     7 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                       seq_1</span></span>
<span><span class="co">#&gt;     [2]     7 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_2</span></span>
<span><span class="co">#&gt;     [3]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                       seq_3</span></span>
<span><span class="co">#&gt;     [4]     7 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_4</span></span>
<span><span class="co">#&gt;     [5]     7 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                       seq_5</span></span>
<span><span class="co">#&gt;     ...   ... ...</span></span>
<span><span class="co">#&gt; [56891]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span>                                       seq_56891</span></span>
<span><span class="co">#&gt; [56892]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_56892</span></span>
<span><span class="co">#&gt; [56893]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_56893</span></span>
<span><span class="co">#&gt; [56894]     7 <span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                       seq_56894</span></span>
<span><span class="co">#&gt; [56895]     7 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span>                                       seq_56895</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="error-correction">Error correction<a class="anchor" aria-label="anchor" href="#error-correction"></a>
</h3>
<p><code>posDemux</code> handles two types reads filtering:</p>
<ul>
<li>Retaining only the reads being matched with a barcode combination
after error correction.</li>
<li>Removing barcode combinations which have few reads as these are
usually artifacts and are not assumed to cover the transcripts of an
entire cell.</li>
</ul>
<p>This subsection consider the first type of filtering. During
demultiplexing,the barcode with the smallest Hamming distance to the
read is chosen and the number of mismatches is recorded. If there are
multiple barcodes being equal in distance to the query, one of these
barcodes is chosen. When there are no mismatches for any of the barcode
sets, we obviously want to keep this read. When one or more barcode has
mismatches, <code>posDemux</code> allows for error correction. In that
case, we will keep the read if the number of mismatches to each of its
barcodes is below a certain threshold.</p>
<p>In our example, the smallest Hamming distance between two different
barcodes is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>3</mn><annotation encoding="application/x-tex">3</annotation></semantics></math>
for all barcode sets. Given
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>,
the number of barcode sequencing errors which can be reliably corrected
in a barcode set, the minimum distance required between two distinct
barcodes needs to be at least
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><mi>n</mi><mo>+</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">2n+1</annotation></semantics></math>.
This means that in our case, we can correct for exactly one sequencing
error:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_demultiplex_res.html">filter_demultiplex_res</a></span><span class="op">(</span></span>
<span>    <span class="va">demultiplex_res</span>,</span>
<span>    allowed_mismatches <span class="op">=</span> <span class="fl">1L</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We will now inspect the summary of the demultiplexing and
filtering:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered_res</span><span class="op">$</span><span class="va">summary_res</span></span>
<span><span class="co">#&gt; Total number of reads: 56895 </span></span>
<span><span class="co">#&gt; Number of reads failing to demultiplex: 4989 (8.77%) </span></span>
<span><span class="co">#&gt; Observed number of unique barcode combinations:978 </span></span>
<span><span class="co">#&gt; Number of possible barcode combinations: 884736 </span></span>
<span><span class="co">#&gt; Estimated number of features: 978.5 </span></span>
<span><span class="co">#&gt; Observed feature to barcode ratio: 0.001105 </span></span>
<span><span class="co">#&gt; Corrected feature to barcode ratio: 0.001106 </span></span>
<span><span class="co">#&gt; Estimated number of observed barcode combinations</span></span>
<span><span class="co">#&gt; corresponding to more than one feature: 0.5 (0.06%) </span></span>
<span><span class="co">#&gt; Number of barcode sets: 3 </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span></span>
<span><span class="co">#&gt; Barcode set: bc3 </span></span>
<span><span class="co">#&gt; Barcode width: 7 </span></span>
<span><span class="co">#&gt; Number of possible barcodes: 96 </span></span>
<span><span class="co">#&gt; Number of allowed mismatches: 1 </span></span>
<span><span class="co">#&gt; Number of reads with 0 mismatches: 51923 (91.26%) </span></span>
<span><span class="co">#&gt; Number of reads with 1 mismatches: 671 (1.18%) </span></span>
<span><span class="co">#&gt; Number of reads above mismatch threshold: 4301 (7.56%) </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span></span>
<span><span class="co">#&gt; Barcode set: bc2 </span></span>
<span><span class="co">#&gt; Barcode width: 7 </span></span>
<span><span class="co">#&gt; Number of possible barcodes: 96 </span></span>
<span><span class="co">#&gt; Number of allowed mismatches: 1 </span></span>
<span><span class="co">#&gt; Number of reads with 0 mismatches: 51918 (91.25%) </span></span>
<span><span class="co">#&gt; Number of reads with 1 mismatches: 567 (1%) </span></span>
<span><span class="co">#&gt; Number of reads above mismatch threshold: 4410 (7.75%) </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span></span>
<span><span class="co">#&gt; Barcode set: bc1 </span></span>
<span><span class="co">#&gt; Barcode width: 7 </span></span>
<span><span class="co">#&gt; Number of possible barcodes: 96 </span></span>
<span><span class="co">#&gt; Number of allowed mismatches: 1 </span></span>
<span><span class="co">#&gt; Number of reads with 0 mismatches: 51930 (91.27%) </span></span>
<span><span class="co">#&gt; Number of reads with 1 mismatches: 642 (1.13%) </span></span>
<span><span class="co">#&gt; Number of reads above mismatch threshold: 4323 (7.6%) </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span></span></code></pre></div>
<p>Now, the results in <code>filtered_res$demultiplex_res</code> only
shows the reads either without mismatches or where error correction
could be applied:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">assigned_barcodes</span><span class="op">)</span></span>
<span><span class="co">#&gt;       bc3      bc2      bc1     </span></span>
<span><span class="co">#&gt; seq_1 "bc3_57" "bc2_51" "bc1_94"</span></span>
<span><span class="co">#&gt; seq_2 "bc3_72" "bc2_94" "bc1_95"</span></span>
<span><span class="co">#&gt; seq_4 "bc3_85" "bc2_81" "bc1_17"</span></span>
<span><span class="co">#&gt; seq_5 "bc3_49" "bc2_64" "bc1_77"</span></span>
<span><span class="co">#&gt; seq_6 "bc3_3"  "bc2_45" "bc1_33"</span></span>
<span><span class="co">#&gt; seq_7 "bc3_17" "bc2_57" "bc1_71"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">mismatches</span><span class="op">)</span></span>
<span><span class="co">#&gt;       bc3 bc2 bc1</span></span>
<span><span class="co">#&gt; seq_1   0   0   0</span></span>
<span><span class="co">#&gt; seq_2   0   0   0</span></span>
<span><span class="co">#&gt; seq_4   0   0   0</span></span>
<span><span class="co">#&gt; seq_5   0   0   0</span></span>
<span><span class="co">#&gt; seq_6   0   0   0</span></span>
<span><span class="co">#&gt; seq_7   0   0   0</span></span></code></pre></div>
<p>We can also see which reads are retained:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">filtered_res</span><span class="op">$</span><span class="va">retained</span><span class="op">)</span></span>
<span><span class="co">#&gt; seq_1 seq_2 seq_3 seq_4 seq_5 seq_6 </span></span>
<span><span class="co">#&gt;  TRUE  TRUE FALSE  TRUE  TRUE  TRUE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filtering-by-barcode-frequency">Filtering by barcode frequency<a class="anchor" aria-label="anchor" href="#filtering-by-barcode-frequency"></a>
</h3>
<p>Now as we have conducted the filtering based on barcode mismatches,
we proceed with the second kind of filtering. It utilizes the Knee
method <a name="cite-Macosko2015"></a>[<a href="https://www.sciencedirect.com/science/article/pii/S0092867415005498" class="external-link">6</a>;
<a href="https://www.nature.com/articles/s41564-020-0729-6" class="external-link">4</a>] where
the barcode combinations are arranged in descending order on frequency
and the most abundant ones are selected. The cutoff is set such that we
achieve a saturation where most reads are covered, yet the remaining
barcode combinations not covered all have a small number of reads.</p>
<p>We first create a frequency table of the barcodes is question:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">freq_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_freq_table.html">create_freq_table</a></span><span class="op">(</span></span>
<span>    <span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">assigned_barcodes</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">freq_table</span><span class="op">)</span></span>
<span><span class="co">#&gt;      bc3    bc2    bc1 frequency cumulative_frequency    fraction</span></span>
<span><span class="co">#&gt; 1 bc3_37  bc2_4 bc1_30       270                  270 0.005201711</span></span>
<span><span class="co">#&gt; 2 bc3_45 bc2_36 bc1_65       246                  516 0.004739336</span></span>
<span><span class="co">#&gt; 3 bc3_25 bc2_34 bc1_25       235                  751 0.004527415</span></span>
<span><span class="co">#&gt; 4 bc3_40 bc2_62 bc1_42       221                  972 0.004257697</span></span>
<span><span class="co">#&gt; 5 bc3_69 bc2_95 bc1_37       219                 1191 0.004219165</span></span>
<span><span class="co">#&gt; 6 bc3_29 bc2_90 bc1_40       217                 1408 0.004180634</span></span>
<span><span class="co">#&gt;   cumulative_fraction</span></span>
<span><span class="co">#&gt; 1         0.005201711</span></span>
<span><span class="co">#&gt; 2         0.009941047</span></span>
<span><span class="co">#&gt; 3         0.014468462</span></span>
<span><span class="co">#&gt; 4         0.018726159</span></span>
<span><span class="co">#&gt; 5         0.022945324</span></span>
<span><span class="co">#&gt; 6         0.027125958</span></span></code></pre></div>
<p>This package contains an interaction Shiny application for selecting
the number of barcodes. It can be run by:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/interactive_bc_cutoff.html">interactive_bc_cutoff</a></span><span class="op">(</span><span class="va">freq_table</span><span class="op">)</span></span></code></pre></div>
<p>If you are working outside RStudio, you may have to run</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/interactive_bc_cutoff.html">interactive_bc_cutoff</a></span><span class="op">(</span><span class="va">freq_table</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">app</span>, launch.browser <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>and opening the resulting link inside the browser. If you are working
on a headless system, you may consider:</p>
<ul>
<li>Using RStudio Server</li>
<li>Opening a reverse ssh tunnel for the Shiny server</li>
<li>Copy the frequency table onto your local machine and run the
application there</li>
</ul>
<p>If you run the application, you will see that keeping approximately
500 barcodes is the optimal choice<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;In case you wonder why the best choice is keeping half
the barcodes: Remember that the data is synthetic, so making half of the
barcode combinations artifacts with low numbers of reads, was a design
decision.&lt;/p&gt;"><sup>1</sup></a>. We can also calculate the frequency of the
least abundant barcode combination being included.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bc_cutoff</span> <span class="op">&lt;-</span> <span class="fl">500L</span></span>
<span><span class="va">freq_cutoff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bc_to_freq_cutoff.html">bc_to_freq_cutoff</a></span><span class="op">(</span><span class="va">freq_table</span>, <span class="va">bc_cutoff</span><span class="op">)</span></span>
<span><span class="va">freq_cutoff</span></span>
<span><span class="co">#&gt; [1] 8</span></span></code></pre></div>
<p>We can convert the cutoff the other way as well, but it is in general
not an exact inverse as multiple barcodes can have the same
frequency:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reconstrued_bc_cutoff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bc_to_freq_cutoff.html">freq_to_bc_cutoff</a></span><span class="op">(</span></span>
<span>    <span class="va">freq_table</span>,</span>
<span>    <span class="va">freq_cutoff</span></span>
<span><span class="op">)</span></span>
<span><span class="va">reconstrued_bc_cutoff</span></span>
<span><span class="co">#&gt; [1] 505</span></span></code></pre></div>
<p>The cutoff can be illustrated by the Knee plot as in the interactive
application:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/freq_plot.html">knee_plot</a></span><span class="op">(</span>freq_table <span class="op">=</span> <span class="va">freq_table</span>, cutoff <span class="op">=</span> <span class="va">bc_cutoff</span><span class="op">)</span></span></code></pre></div>
<p><img src="demultiplexing_files/figure-html/knee_plot-1.png" class="r-plt" width="700"></p>
<p>We can also illustrate this filtering by looking at the distribution
of barcode frequencies:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Since the cutoff lines of the plot are provided by the literal x-coordinate,</span></span>
<span><span class="co"># we must use the frequency cutoff</span></span>
<span><span class="fu"><a href="../reference/freq_plot.html">freq_plot</a></span><span class="op">(</span><span class="va">freq_table</span>,</span>
<span>    cutoff <span class="op">=</span> <span class="va">freq_cutoff</span>, type <span class="op">=</span> <span class="st">"density"</span>,</span>
<span>    log_scale_x <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="demultiplexing_files/figure-html/freq_plot-1.png" class="r-plt" width="700"></p>
<p>Often the frequency plot gets easier to interpret when we scale the
y-axis by the number of reads, hence creating a mass plot of where the
reads are distributed based on their respective barcodes:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Since the cutoff lines of the plot are provided by the literal x-coordinate,</span></span>
<span><span class="co"># we must use the frequency cutoff</span></span>
<span><span class="fu"><a href="../reference/freq_plot.html">freq_plot</a></span><span class="op">(</span><span class="va">freq_table</span>,</span>
<span>    cutoff <span class="op">=</span> <span class="va">freq_cutoff</span>, type <span class="op">=</span> <span class="st">"density"</span>,</span>
<span>    log_scale_x <span class="op">=</span> <span class="cn">TRUE</span>, scale_by_reads <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="demultiplexing_files/figure-html/mass_plot-1.png" class="r-plt" width="700"></p>
</div>
<div class="section level3">
<h3 id="exporting-results">Exporting results<a class="anchor" aria-label="anchor" href="#exporting-results"></a>
</h3>
<p>Now as we know how many barcodes to keep, we can extract the desired
results and write it to a table.</p>
<p>We start with the frequency table. Since it is already sorted in
descending order of frequency, we simply take the top entries in the
table:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_freq_table</span> <span class="op">&lt;-</span> <span class="va">freq_table</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">bc_cutoff</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>For finding which reads correspond to these barcodes, we do the
following:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assigned_barcodes</span> <span class="op">&lt;-</span> <span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">assigned_barcodes</span></span>
<span><span class="va">read_in_selection</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/row_match.html">row_match</a></span><span class="op">(</span><span class="va">assigned_barcodes</span>, <span class="va">selected_freq_table</span><span class="op">)</span></span>
<span><span class="va">selected_assigned_barcodes</span> <span class="op">&lt;-</span> <span class="va">assigned_barcodes</span><span class="op">[</span><span class="va">read_in_selection</span>, <span class="op">]</span></span></code></pre></div>
<p>We continue with the UMIs which we convert to a character vector
(remember that the payload and assigned barcodes are aligned in
order):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assigned_UMI</span> <span class="op">&lt;-</span> <span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">payload</span><span class="op">$</span><span class="va">UMI</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">selected_assigned_UMI</span> <span class="op">&lt;-</span> <span class="va">assigned_UMI</span><span class="op">[</span><span class="va">read_in_selection</span><span class="op">]</span></span></code></pre></div>
<p>With all of this done, we can make a data frame containing the UMI,
read identifier, and barcode assignments:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">selected_assigned_barcodes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        read <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">selected_assigned_barcodes</span><span class="op">)</span>,</span>
<span>        UMI <span class="op">=</span> <span class="va">selected_assigned_UMI</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># Ensures the columns appears in the desired order</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">read</span>, <span class="va">UMI</span>, <span class="va">bc3</span>, <span class="va">bc2</span>, <span class="va">bc1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res_table</span><span class="op">)</span></span>
<span><span class="co">#&gt;        read     UMI    bc3    bc2    bc1</span></span>
<span><span class="co">#&gt; seq_1 seq_1 GCCTAAC bc3_57 bc2_51 bc1_94</span></span>
<span><span class="co">#&gt; seq_2 seq_2 CCAAGCG bc3_72 bc2_94 bc1_95</span></span>
<span><span class="co">#&gt; seq_4 seq_4 CCTAACG bc3_85 bc2_81 bc1_17</span></span>
<span><span class="co">#&gt; seq_5 seq_5 GCTCGTC bc3_49 bc2_64 bc1_77</span></span>
<span><span class="co">#&gt; seq_6 seq_6 TGGAGAA  bc3_3 bc2_45 bc1_33</span></span>
<span><span class="co">#&gt; seq_7 seq_7 ACTTCGA bc3_17 bc2_57 bc1_71</span></span></code></pre></div>
<p>Finally, we can write the table to file. The following is the
suggested way for achieving the desired formatting:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"barcode_table"</span>, fileext <span class="op">=</span> <span class="st">".txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">res_table</span>, <span class="va">file</span>,</span>
<span>    row.names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col.names <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, eol <span class="op">=</span> <span class="st">"\n"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.23/posDemux" class="external-link">posDemux</a></em>
package<br>
was made possible thanks to:</p>
<ul>
<li>R</li>
<li><em><a href="https://bioconductor.org/packages/3.23/BiocStyle" class="external-link">BiocStyle</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=sessioninfo" class="external-link">sessioninfo</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=testthat" class="external-link">testthat</a></em></li>
</ul>
<p>This package was developed using <em><a href="https://bioconductor.org/packages/3.23/biocthis" class="external-link">biocthis</a></em>.</p>
<p><code>R</code> session information:</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R Under development (unstable) (2025-12-12 r89163)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2025-12-15</span></span>
<span><span class="co">#&gt;  pandoc   3.1.11 @ /opt/hostedtoolcache/pandoc/3.1.11/x64/ (via rmarkdown)</span></span>
<span><span class="co">#&gt;  quarto   NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package     </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version  </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  assertthat     0.2.1     <span style="color: #555555;">2019-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  backports      1.5.0     <span style="color: #555555;">2024-05-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bibtex         0.5.1     <span style="color: #555555;">2023-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  BiocGenerics * 0.57.0    <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocManager    1.30.27   <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  BiocStyle    * 2.39.0    <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  Biostrings   * 2.79.2    <span style="color: #555555;">2025-11-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  bookdown       0.46      <span style="color: #555555;">2025-12-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib          0.9.0     <span style="color: #555555;">2025-01-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem         1.1.0     <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli            3.6.5     <span style="color: #555555;">2025-04-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  crayon         1.5.3     <span style="color: #555555;">2024-06-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc           1.4.3     <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest         0.6.39    <span style="color: #555555;">2025-11-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dplyr          1.1.4     <span style="color: #555555;">2023-11-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate       1.0.5     <span style="color: #555555;">2025-08-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  farver         2.1.2     <span style="color: #555555;">2024-05-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap        1.2.0     <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs             1.6.6     <span style="color: #555555;">2025-04-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics     * 0.1.4     <span style="color: #555555;">2025-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggplot2        4.0.1     <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue           1.8.0     <span style="color: #555555;">2024-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable         0.3.6     <span style="color: #555555;">2024-10-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools      0.5.9     <span style="color: #555555;">2025-12-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets    1.6.4     <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httpuv         1.6.16    <span style="color: #555555;">2025-04-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httr           1.4.7     <span style="color: #555555;">2023-08-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  IRanges      * 2.45.0    <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  jquerylib      0.1.4     <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite       2.0.0     <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr          1.50      <span style="color: #555555;">2025-03-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  labeling       0.4.3     <span style="color: #555555;">2023-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  later          1.4.4     <span style="color: #555555;">2025-08-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lifecycle      1.0.4     <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lubridate      1.9.4     <span style="color: #555555;">2024-12-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr     * 2.0.4     <span style="color: #555555;">2025-09-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mime           0.13      <span style="color: #555555;">2025-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  otel           0.2.0     <span style="color: #555555;">2025-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pillar         1.11.1    <span style="color: #555555;">2025-09-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig      2.0.3     <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown        2.2.0     <span style="color: #555555;">2025-11-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  plyr           1.8.9     <span style="color: #555555;">2023-10-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  posDemux     * 0.99.8    <span style="color: #555555;">2025-12-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  promises       1.5.0     <span style="color: #555555;">2025-11-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr        * 1.2.0     <span style="color: #555555;">2025-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R6             2.6.1     <span style="color: #555555;">2025-02-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg           1.5.0     <span style="color: #555555;">2025-09-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RColorBrewer   1.1-3     <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp           1.1.0.8.1 <span style="color: #555555;">2025-12-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RefManageR   * 1.4.0     <span style="color: #555555;">2022-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rlang          1.1.6     <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown      2.30      <span style="color: #555555;">2025-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  S4Vectors    * 0.49.0    <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  S7             0.2.1     <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sass           0.4.10    <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales         1.4.0     <span style="color: #555555;">2025-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Seqinfo      * 1.1.0     <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  sessioninfo  * 1.2.3     <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shiny          1.12.1    <span style="color: #555555;">2025-12-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringi        1.8.7     <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr        1.6.0     <span style="color: #555555;">2025-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  systemfonts    1.3.1     <span style="color: #555555;">2025-10-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping    1.0.4     <span style="color: #555555;">2025-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble         3.3.0     <span style="color: #555555;">2025-06-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect     1.2.1     <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  timechange     0.3.0     <span style="color: #555555;">2024-01-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs          0.6.5     <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr          3.0.2     <span style="color: #555555;">2024-10-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun           0.54      <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xml2           1.5.1     <span style="color: #555555;">2025-12-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xtable         1.8-4     <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  XVector      * 0.51.0    <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  yaml           2.3.12    <span style="color: #555555;">2025-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/devel/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/devel/lib/R/library</span></span></span>
<span><span class="co">#&gt;  <span style="color: #BBBBBB; background-color: #BB0000;">*</span> ── Packages attached to the search path.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.23/BiocStyle" class="external-link">BiocStyle</a></em>
with <em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em> and
<em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em>
running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em>
.</p>
<p>
<a id="bib-Rosenberg2018"></a><a href="#cite-Rosenberg2018">[1]</a><cite>
A. B. Rosenberg, C. M. Roco, R. A. Muscat, et al. “Single-cell profiling
of the developing mouse brain and spinal cord with split-pool
barcoding”. In: <em>Science</em> 360.6385 (Apr. 2018), pp. 176–182. DOI:
<a href="https://doi.org/10.1126/science.aam8999" class="external-link">10.1126/science.aam8999</a>.
URL:
<a href="https://www.science.org/doi/10.1126/science.aam8999" class="external-link">https://www.science.org/doi/10.1126/science.aam8999</a>
(visited on 01/23/2025).</cite>
</p>
<p>
<a id="bib-Kuijpers2024"></a><a href="#cite-Kuijpers2024">[2]</a><cite>
L. Kuijpers, B. Hornung, M. C. G. N. van den Hout - van Vroonhoven, et
al. “Split Pool Ligation-based Single-cell Transcriptome sequencing
(SPLiT-seq) data processing pipeline comparison”. In: <em>BMC
Genomics</em> 25.1 (Apr. 2024), p. 361. ISSN: 1471-2164. DOI:
<a href="https://doi.org/10.1186/s12864-024-10285-3" class="external-link">10.1186/s12864-024-10285-3</a>.
URL:
<a href="https://doi.org/10.1186/s12864-024-10285-3" class="external-link">https://doi.org/10.1186/s12864-024-10285-3</a>
(visited on 01/08/2025).</cite>
</p>
<p>
<a id="bib-Ma2023"></a><a href="#cite-Ma2023">[3]</a><cite> P. Ma, H. M.
Amemiya, L. L. He, et al. “Bacterial droplet-based single-cell RNA-seq
reveals antibiotic-associated heterogeneous cellular states”. In:
<em>Cell</em> 186.4 (Feb. 2023), pp. 877–891.e14. ISSN: 0092-8674. DOI:
<a href="https://doi.org/10.1016/j.cell.2023.01.002" class="external-link">10.1016/j.cell.2023.01.002</a>.
URL:
<a href="https://www.sciencedirect.com/science/article/pii/S0092867423000028" class="external-link">https://www.sciencedirect.com/science/article/pii/S0092867423000028</a>
(visited on 01/31/2025).</cite>
</p>
<p>
<a id="bib-Blattman2020"></a><a href="#cite-Blattman2020">[4]</a><cite>
S. B. Blattman, W. Jiang, P. Oikonomou, et al. “Prokaryotic single-cell
RNA sequencing by in situ combinatorial indexing”. En. In: <em>Nature
Microbiology</em> 5.10 (Oct. 2020), pp. 1192–1201. ISSN: 2058-5276. DOI:
<a href="https://doi.org/10.1038/s41564-020-0729-6" class="external-link">10.1038/s41564-020-0729-6</a>.
URL:
<a href="https://www.nature.com/articles/s41564-020-0729-6" class="external-link">https://www.nature.com/articles/s41564-020-0729-6</a>
(visited on 12/12/2024).</cite>
</p>
<p>
<a id="bib-Blattman2024"></a><a href="#cite-Blattman2024">[5]</a><cite>
S. B. Blattman, W. Jiang, E. R. McGarrigle, et al. “Identification and
genetic dissection of convergent persister cell states”. En. In:
<em>Nature</em> 636.8042 (Dec. 2024), pp. 438–446. ISSN: 1476-4687. DOI:
<a href="https://doi.org/10.1038/s41586-024-08124-2" class="external-link">10.1038/s41586-024-08124-2</a>.
URL:
<a href="https://www.nature.com/articles/s41586-024-08124-2" class="external-link">https://www.nature.com/articles/s41586-024-08124-2</a>
(visited on 01/08/2025).</cite>
</p>
<p>
<a id="bib-Macosko2015"></a><a href="#cite-Macosko2015">[6]</a><cite> E.
Macosko, A. Basu, R. Satija, et al. “Highly Parallel Genome-wide
Expression Profiling of Individual Cells Using Nanoliter Droplets”. In:
<em>Cell</em> 161.5 (2015), pp. 1202-1214. ISSN: 0092-8674. DOI:
<a href="https://doi.org/https://doi.org/10.1016/j.cell.2015.05.002" class="external-link">https://doi.org/10.1016/j.cell.2015.05.002</a>.
URL:
<a href="https://www.sciencedirect.com/science/article/pii/S0092867415005498" class="external-link">https://www.sciencedirect.com/science/article/pii/S0092867415005498</a>.</cite>
</p>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jakob Peder Pettersen, Centre for new antibacterial strategies (CANS).</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
