<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Streaming data into the demultiplexer • posDemux</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Streaming data into the demultiplexer">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">posDemux</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.10</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/posDemux.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/streaming.html">Streaming data into the demultiplexer</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/yaccos/posDemux/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Streaming data into the demultiplexer</h1>
                        <h4 data-toc-skip class="author">Jakob Peder
Pettersen</h4>
            <address class="author_afil">
      UiT, the Artic University of Norway, Department of Computer
Science<br><a class="author_email" href="mailto:#"></a><a href="mailto:jakobpeder.pettersen@gmail.com" class="email">jakobpeder.pettersen@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">21 January 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/yaccos/posDemux/blob/main/vignettes/streaming.Rmd" class="external-link"><code>vignettes/streaming.Rmd</code></a></small>
      <div class="d-none name"><code>streaming.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="why-streaming">Why streaming?<a class="anchor" aria-label="anchor" href="#why-streaming"></a>
</h2>
<p>Given that sequence files can consist of gigabytes of data, storing
everything in memory at once can prove to be too much for the computer.
Fortunately, for demultiplexing of barcodes, we can demultiplex the
barcodes independently of each other. In theory, we could read the
entries from a FASTQ file one by one, and for each of them call the
demultiplexer and output the barcode assignment to a file. However, due
to overhead in file operations and the <code>R</code> interpreter, the
most efficient option is to process the files in chunks of a
predetermined size. If you have not read the main package vignette yet
(<code><a href="../articles/posDemux.html">vignette("posDemux")</a></code>), please do so before proceeding.</p>
<div class="section level3">
<h3 id="which-chunk-size-to-use">Which chunk size to use?<a class="anchor" aria-label="anchor" href="#which-chunk-size-to-use"></a>
</h3>
<p>The chunk size should be large enough for the constant overhead of
initiating a file read and interpreting the required <code>R</code>
instructions to be neglectable. Still, the chunk size should be small
enough to only make a modest memory footprint. In practice, with modern
computers having gigabytes of RAM, we find a chunk size of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>2</mn><mo>⋅</mo><msup><mn>10</mn><mn>5</mn></msup></mrow><annotation encoding="application/x-tex">2\cdot10^5</annotation></semantics></math>
reads to be a reasonable choice. The example datasets used in this
vignette are far too small to require streaming (including a dataset of
realistic size would make the package exceed the maximum package size),
so for purely demonstrational purposes, we choose a chunk size of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mn>10</mn><mn>4</mn></msup><annotation encoding="application/x-tex">10^4</annotation></semantics></math>
reads.</p>
</div>
<div class="section level3">
<h3 id="overview-of-the-streaming-api">Overview of the streaming API<a class="anchor" aria-label="anchor" href="#overview-of-the-streaming-api"></a>
</h3>
<p>The streaming API of <code>posDemux</code> is focused around two
user-supplied functions:</p>
<ul>
<li>The loader: This function is responsible for grabbing a chunk of the
input file of the sequences and returning a <code>XStringSet</code>
object of the sequences. Also, this function determines when to
terminate the streaming, typically done when there are no sequences
left.</li>
<li>The archiver: This function is responsible for writing the results
of the demultiplexer of a chunk to file, typically in the form of a
barcode table.</li>
</ul>
<p>Once the loader and archiver are specified and combined with the
demultiplexer, they can generate the barcode table chunk by chunk.
However, we still want to generate the summary statistics and frequency
table as if all the reads were demultiplexed simultaneously. If the
chunks are demultiplexed independently,
<code><a href="../reference/create_summary_res.html">create_summary_res()</a></code> and <code><a href="../reference/create_freq_table.html">create_freq_table()</a></code>
don’t work as intended. That is where the core function of the streaming
API; <code><a href="../reference/streaming_demultiplex.html">streaming_demultiplex()</a></code> comes into play. It accepts
the loader and the archiver as callback functions, communicates with the
demultiplexer, and does the bookkeeping to construct the summary
statistics and frequency table. Hence, for each chunk, the loader
provides the sequences to demultiplex, the streaming logic feeds these
reads into the demultiplexer, updates is record of observed barcodes and
provides the results to the archiver which writes to the barcode table.
This process is illustrated in Figure @ref(fig:structure).</p>
<p>For communicating between the loader and archiver, a user-defined
state object is passed through <code><a href="../reference/streaming_demultiplex.html">streaming_demultiplex()</a></code>. The
initial value of this object is provided to
<code><a href="../reference/streaming_demultiplex.html">streaming_demultiplex()</a></code> as the argument
<code>state_init</code>.</p>
<div class="figure">
<img src="streaming_structure.svg" class="r-plt" alt="Components of the streaming API" width="100%"><p class="caption">
Components of the streaming API
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="streaming-using-default-callbacks">Streaming using default callbacks<a class="anchor" aria-label="anchor" href="#streaming-using-default-callbacks"></a>
</h2>
<p>Although the option of specifying the loader and archiver gives great
flexibility, it does so at the cost of convenience as the loader and
archiver can be challenging to write. For this reason,
<code>posDemux</code> provides a default set of streaming callbacks
available through <code><a href="../reference/streaming_callbacks.html">streaming_callbacks()</a></code>. Actually, this is
a function factory, which returns a list of the following elements:</p>
<ul>
<li>
<code>state_init</code>: The initial state to provide to
<code><a href="../reference/streaming_demultiplex.html">streaming_demultiplex()</a></code>
</li>
<li>
<code>loader</code>: Loader function</li>
<li>
<code>archiver</code>: Archiver function</li>
</ul>
<div class="section level3">
<h3 id="assumptions">Assumptions<a class="anchor" aria-label="anchor" href="#assumptions"></a>
</h3>
<p>The callbacks are designed to be useful for a typical use case
where:</p>
<ul>
<li>The reads are contained as a single FASTQ file on disk.</li>
<li>The barcode table contains columns with the read names
(<code>read</code>), sequences of all payloads (eg.<code>UMI</code>) and
the barcode assignments (eg. <code>BC3</code>, <code>BC2</code>,
<code>BC1</code>).</li>
<li>Read names with spaces are truncated by the loader such that only
the part of the read name before the space is kept. This is done because
the information preceding the space is typically enough to ambiguously
identify the read, while the information after the space usually
contains irrelevant auxiliary information.</li>
<li>Progress of how many reads have been processed can optionally be
displayed by the archiver for each chunk.</li>
</ul>
<p>If this does not fill your needs, please refer to the section <a href="#streaming-using-custom-callbacks">Streaming using custom
callbacks</a>.</p>
</div>
<div class="section level3">
<h3 id="obtaining-the-callbacks">Obtaining the callbacks<a class="anchor" aria-label="anchor" href="#obtaining-the-callbacks"></a>
</h3>
<p>As said, <code><a href="../reference/streaming_callbacks.html">streaming_callbacks()</a></code> is a function factory, we
must specify the file path to the FASTQ input file, and the barcode
table output file for it to generate useful callbacks. In addition, the
chunk size is regulated through the callbacks. We use the same dataset
as in the main package vignette
(<code>vignette("demultiplexing")</code>) and obtain the callbacks as
follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yaccos/posDemux" class="external-link">posDemux</a></span><span class="op">)</span></span>
<span><span class="va">input_fastq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    <span class="st">"PETRI-seq_forward_reads.fq.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"posDemux"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">output_barcode_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span></span>
<span>    pattern <span class="op">=</span> <span class="st">"barcode_table"</span>,</span>
<span>    fileext <span class="op">=</span> <span class="st">".txt"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">default_callbacks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/streaming_callbacks.html">streaming_callbacks</a></span><span class="op">(</span></span>
<span>    input_file <span class="op">=</span> <span class="va">input_fastq</span>,</span>
<span>    output_table_file <span class="op">=</span> <span class="va">output_barcode_table</span>,</span>
<span>    chunk_size <span class="op">=</span> <span class="fl">1e+4</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-the-demultiplexing">Running the demultiplexing<a class="anchor" aria-label="anchor" href="#running-the-demultiplexing"></a>
</h3>
<p>With the callbacks in place, we specify the barcodes, and the
segments and feed them into the streaming framework:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">barcode_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata/PETRI-seq_barcodes"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>bc1 <span class="op">=</span> <span class="st">"bc1.fa"</span>, bc2 <span class="op">=</span> <span class="st">"bc2.fa"</span>, bc3 <span class="op">=</span> <span class="st">"bc3.fa"</span><span class="op">)</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"posDemux"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">barcode_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"bc"</span>, <span class="fl">1L</span><span class="op">:</span><span class="fl">3L</span><span class="op">)</span></span>
<span><span class="va">barcode_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">barcode_files</span>, <span class="va">readDNAStringSet</span><span class="op">)</span></span>
<span><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="va">barcode_index</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bc3"</span>, <span class="st">"bc2"</span>, <span class="st">"bc1"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sequence_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>UMI <span class="op">=</span> <span class="st">"P"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"A"</span><span class="op">)</span></span>
<span><span class="va">segment_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7L</span>, <span class="fl">7L</span>, <span class="fl">15L</span>, <span class="fl">7L</span>, <span class="fl">14L</span>, <span class="fl">7L</span>, <span class="cn">NA_integer_</span><span class="op">)</span></span>
<span><span class="va">streaming_summary_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/streaming_demultiplex.html">streaming_demultiplex</a></span><span class="op">(</span></span>
<span>    state_init <span class="op">=</span> <span class="va">default_callbacks</span><span class="op">$</span><span class="va">state_init</span>,</span>
<span>    loader <span class="op">=</span> <span class="va">default_callbacks</span><span class="op">$</span><span class="va">loader</span>,</span>
<span>    archiver <span class="op">=</span> <span class="va">default_callbacks</span><span class="op">$</span><span class="va">archiver</span>,</span>
<span>    barcodes <span class="op">=</span> <span class="va">barcodes</span>,</span>
<span>    allowed_mismatches <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>    segments <span class="op">=</span> <span class="va">sequence_annotation</span>,</span>
<span>    segment_lengths <span class="op">=</span> <span class="va">segment_lengths</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:22 2026 =&gt; Initializing FASTQ stream and output table</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:22 2026 =&gt; Streaming FASTQ input file /home/runner/work/_temp/Library/posDemux/extdata/PETRI-seq_forward_reads.fq.gz</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:28 2026 =&gt; Processed 10000 reads, successfully demultiplexed 9117 reads so far...</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:28 2026 =&gt; Processed 20000 reads, successfully demultiplexed 18259 reads so far...</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:28 2026 =&gt; Processed 30000 reads, successfully demultiplexed 27362 reads so far...</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:28 2026 =&gt; Processed 40000 reads, successfully demultiplexed 36487 reads so far...</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:28 2026 =&gt; Processed 50000 reads, successfully demultiplexed 45629 reads so far...</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:28 2026 =&gt; Processed 56895 reads, successfully demultiplexed 51906 reads so far...</span></span>
<span><span class="co">#&gt; Wed Jan 21 16:11:28 2026 =&gt; Done demultiplexing</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="displaying-the-result-from-streaming">Displaying the result from streaming<a class="anchor" aria-label="anchor" href="#displaying-the-result-from-streaming"></a>
</h3>
<p>From the return value of <code><a href="../reference/streaming_demultiplex.html">streaming_demultiplex()</a></code>, we
obtain the frequency table</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">streaming_summary_res</span><span class="op">$</span><span class="va">freq_table</span><span class="op">)</span></span>
<span><span class="co">#&gt;      bc3    bc2    bc1 frequency cumulative_frequency    fraction</span></span>
<span><span class="co">#&gt; 1 bc3_37  bc2_4 bc1_30       270                  270 0.005201711</span></span>
<span><span class="co">#&gt; 2 bc3_45 bc2_36 bc1_65       246                  516 0.004739336</span></span>
<span><span class="co">#&gt; 3 bc3_25 bc2_34 bc1_25       235                  751 0.004527415</span></span>
<span><span class="co">#&gt; 4 bc3_40 bc2_62 bc1_42       221                  972 0.004257697</span></span>
<span><span class="co">#&gt; 5 bc3_69 bc2_95 bc1_37       219                 1191 0.004219165</span></span>
<span><span class="co">#&gt; 6 bc3_56 bc2_94 bc1_29       217                 1408 0.004180634</span></span>
<span><span class="co">#&gt;   cumulative_fraction</span></span>
<span><span class="co">#&gt; 1         0.005201711</span></span>
<span><span class="co">#&gt; 2         0.009941047</span></span>
<span><span class="co">#&gt; 3         0.014468462</span></span>
<span><span class="co">#&gt; 4         0.018726159</span></span>
<span><span class="co">#&gt; 5         0.022945324</span></span>
<span><span class="co">#&gt; 6         0.027125958</span></span></code></pre></div>
<p>and summary printout:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">streaming_summary_res</span><span class="op">$</span><span class="va">summary_res</span></span>
<span><span class="co">#&gt; Total number of reads: 56895 </span></span>
<span><span class="co">#&gt; Number of reads failing to demultiplex: 4989 (8.77%) </span></span>
<span><span class="co">#&gt; Observed number of unique barcode combinations:978 </span></span>
<span><span class="co">#&gt; Number of possible barcode combinations: 884736 </span></span>
<span><span class="co">#&gt; Estimated number of features: 978.5 </span></span>
<span><span class="co">#&gt; Observed feature to barcode ratio: 0.001105 </span></span>
<span><span class="co">#&gt; Corrected feature to barcode ratio: 0.001106 </span></span>
<span><span class="co">#&gt; Estimated number of observed barcode combinations</span></span>
<span><span class="co">#&gt; corresponding to more than one feature: 0.5 (0.06%) </span></span>
<span><span class="co">#&gt; Number of barcode sets: 3 </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span></span>
<span><span class="co">#&gt; Barcode set: bc3 </span></span>
<span><span class="co">#&gt; Barcode width: 7 </span></span>
<span><span class="co">#&gt; Number of possible barcodes: 96 </span></span>
<span><span class="co">#&gt; Number of allowed mismatches: 1 </span></span>
<span><span class="co">#&gt; Number of reads with 0 mismatches: 51923 (91.26%) </span></span>
<span><span class="co">#&gt; Number of reads with 1 mismatches: 671 (1.18%) </span></span>
<span><span class="co">#&gt; Number of reads above mismatch threshold: 4301 (7.56%) </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span></span>
<span><span class="co">#&gt; Barcode set: bc2 </span></span>
<span><span class="co">#&gt; Barcode width: 7 </span></span>
<span><span class="co">#&gt; Number of possible barcodes: 96 </span></span>
<span><span class="co">#&gt; Number of allowed mismatches: 1 </span></span>
<span><span class="co">#&gt; Number of reads with 0 mismatches: 51918 (91.25%) </span></span>
<span><span class="co">#&gt; Number of reads with 1 mismatches: 567 (1%) </span></span>
<span><span class="co">#&gt; Number of reads above mismatch threshold: 4410 (7.75%) </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - </span></span>
<span><span class="co">#&gt; Barcode set: bc1 </span></span>
<span><span class="co">#&gt; Barcode width: 7 </span></span>
<span><span class="co">#&gt; Number of possible barcodes: 96 </span></span>
<span><span class="co">#&gt; Number of allowed mismatches: 1 </span></span>
<span><span class="co">#&gt; Number of reads with 0 mismatches: 51930 (91.27%) </span></span>
<span><span class="co">#&gt; Number of reads with 1 mismatches: 642 (1.13%) </span></span>
<span><span class="co">#&gt; Number of reads above mismatch threshold: 4323 (7.6%) </span></span>
<span><span class="co">#&gt; - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="streaming-the-barcode-table">Streaming the barcode table<a class="anchor" aria-label="anchor" href="#streaming-the-barcode-table"></a>
</h2>
<div class="section level3">
<h3 id="considerations-about-size-of-tables">Considerations about size of tables<a class="anchor" aria-label="anchor" href="#considerations-about-size-of-tables"></a>
</h3>
<p>The frequency table from a realistic PETRI-seq dataset is relatively
small (just a few megabytes) and should therefore not pose any
substantial memory usage problems. On the other hand, the barcode table
will have one row for each read processed, making its memory footprint
closer to the size of the original FASTQ-file. To this end, the default
archiver writes to the barcode table on disk for each chunk, thus
avoiding keeping the table in memory. However, the barcode table will be
required by some downstream step of the pipeline. Typically, this will
be frequency filtering. The good news is that the cutoff determination
process only needs the light frequency table. The bad news is that the
full barcode table is required when determining which reads to keep
after the cutoff is selected.</p>
<p>If we want to avoid heavy load on RAM, we must stream the barcode
table in chunks, determine for each chunk which reads to keep, and
appending the results to another file.</p>
</div>
<div class="section level3">
<h3 id="subsetting-the-frequency-table">Subsetting the frequency table<a class="anchor" aria-label="anchor" href="#subsetting-the-frequency-table"></a>
</h3>
<p>We begin this example by subsetting the frequency table. In order not
to repeat ourselves, we refer to the main package vignette
(<code>vignette("demultiplexing")</code>) for the considerations behind
this process.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">freq_table</span> <span class="op">&lt;-</span> <span class="va">streaming_summary_res</span><span class="op">$</span><span class="va">freq_table</span></span>
<span><span class="va">bc_cutoff</span> <span class="op">&lt;-</span> <span class="fl">500L</span></span>
<span><span class="va">selected_freq_table</span> <span class="op">&lt;-</span> <span class="va">freq_table</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">bc_cutoff</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="writing-to-database">Writing to database<a class="anchor" aria-label="anchor" href="#writing-to-database"></a>
</h3>
<p>We typically want to create a copy of the barcode table including
only those reads which are selected by the barcode frequency and thus
match a row in <code>selected_freq_table</code>. We
<strong>could</strong> use the most direct approach and provide the
selected barcode table in the same file format, just with desired rows
kept. However, this could become impractical for downstream analysis as
using this new table as a lookup would require us to load it all into
memory, causing the same problem one level down. For memory-constrained
environments, we therefore suggest another idea: Dumping the data into a
SQLite database which later can be queried. For simplicity, we paste all
the barcode designations together in one column. We begin by setting up
the database:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dbi.r-dbi.org" class="external-link">DBI</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">output_database_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span></span>
<span>    pattern <span class="op">=</span> <span class="st">"selected_barcode_table"</span>,</span>
<span>    fileext <span class="op">=</span> <span class="st">".sqlite"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">output_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dbi.r-dbi.org/reference/dbConnect.html" class="external-link">dbConnect</a></span><span class="op">(</span><span class="fu">RSQLite</span><span class="fu">::</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">output_database_path</span><span class="op">)</span></span></code></pre></div>
<p>For doing the streaming of the barcode table and combining it with
outputting to the database, we use the <code>chunked</code> package:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chunk_size</span> <span class="op">&lt;-</span> <span class="fl">1e4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/edwindj/chunked" class="external-link">chunked</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/chunked/man/read_chunks.html" class="external-link">read_table_chunkwise</a></span><span class="op">(</span></span>
<span>    file <span class="op">=</span> <span class="va">output_barcode_table</span>,</span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    chunk_size <span class="op">=</span> <span class="va">chunk_size</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># For use within a pipeline, it is more convenient to use inner_join() </span></span>
<span>    <span class="co"># than posDemux::row_match() and it achieves the save result</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">selected_freq_table</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        celltag <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span></span>
<span>            <span class="va">paste</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">barcodes</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">read</span>, <span class="va">UMI</span>, <span class="va">celltag</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/chunked/man/write_chunkwise.html" class="external-link">write_chunkwise</a></span><span class="op">(</span><span class="fu">dbplyr</span><span class="fu">::</span><span class="fu"><a href="https://dbplyr.tidyverse.org/reference/src_dbi.html" class="external-link">src_dbi</a></span><span class="op">(</span><span class="va">output_db</span><span class="op">)</span>, <span class="st">"selected_barcodes"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(bc3, bc2, bc1)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(bc3, bc2, bc1)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(bc3, bc2, bc1)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(bc3, bc2, bc1)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(bc3, bc2, bc1)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(bc3, bc2, bc1)`</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Source:   table&lt;`selected_barcodes`&gt; [?? x 3]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Database: sqlite 3.51.1 [/tmp/RtmpElCrMK/selected_barcode_table40ee399737d6.sqlite]</span></span></span>
<span><span class="co">#&gt;    read   UMI     celltag             </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> seq_1  GCCTAAC bc3_57_bc2_51_bc1_94</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> seq_2  CCAAGCG bc3_72_bc2_94_bc1_95</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> seq_4  CCTAACG bc3_85_bc2_81_bc1_17</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> seq_5  GCTCGTC bc3_49_bc2_64_bc1_77</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> seq_6  TGGAGAA bc3_3_bc2_45_bc1_33 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> seq_7  ACTTCGA bc3_17_bc2_57_bc1_71</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> seq_8  GCCCGCA bc3_69_bc2_65_bc1_49</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> seq_9  CCCATGT bc3_38_bc2_87_bc1_83</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> seq_10 TGCCGGT bc3_76_bc2_36_bc1_85</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> seq_11 CAACTCA bc3_15_bc2_84_bc1_16</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ more rows</span></span></span></code></pre></div>
<p>Finally, we make a index on the barcodes such that lookups are fast
and close the database connection:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbExecute.html" class="external-link">dbExecute</a></span><span class="op">(</span><span class="va">output_db</span>, <span class="st">"CREATE INDEX idx ON selected_barcodes(read)"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbDisconnect.html" class="external-link">dbDisconnect</a></span><span class="op">(</span><span class="va">output_db</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="streaming-using-custom-callbacks">Streaming using custom callbacks<a class="anchor" aria-label="anchor" href="#streaming-using-custom-callbacks"></a>
</h2>
<div class="section level3">
<h3 id="desired-properties">Desired properties<a class="anchor" aria-label="anchor" href="#desired-properties"></a>
</h3>
<p>In cases where the default callbacks are not sufficient, it is
possible for the user to write custom callback functions. For our
example, we will use the same sample dataset as before, but pretend the
reads for some reason come in a FASTA file (which the default loader
does not support). In addition, we want the UMIs to go into its own
FASTA files instead of being a column in the barcode table. Unlike the
default callbacks, we don’t truncate read names.</p>
</div>
<div class="section level3">
<h3 id="the-state-object">The state object<a class="anchor" aria-label="anchor" href="#the-state-object"></a>
</h3>
<p>For progress printouts, we want the state object of the streaming to
contain the total number of reads and the number of reads where
demultiplexing was successful. Additionally, for our specific case, the
loader needs to know where to begin reading the FASTA file for each
chunk. For the final numbers, this is not really necessary because the
required information is provided by the streaming framework itself when
finishing. Thus, we want the state object to contain the following
fields which are both integers:</p>
<ul>
<li><code>total_reads</code></li>
<li><code>demultiplexed_reads</code></li>
</ul>
<p><code>total_reads</code> will be updated by the loader, whereas
<code>demultiplexed_reads</code> will be updated by the archiver at the
end of each chunk. Furthermore, the archiver needs to write the column
headers to the barcode table when first writing to it, so it must know
whether it is working at its first chunk. This gives rise to the logical
flag <code>output_table_initialized</code>. We thus initialize the state
as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_state_init</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    total_reads <span class="op">=</span> <span class="fl">0L</span>, demultiplexed_reads <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>    output_table_initialized <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-loader">The loader<a class="anchor" aria-label="anchor" href="#the-loader"></a>
</h3>
<p>The loader is a function which accepts the state as its sole
parameter and outputs a list with three fields:</p>
<ul>
<li>
<code>state</code>: To be passed into the archiver.</li>
<li>
<code>sequences</code>: The sequences in the chunk as a
<code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">Biostrings::DNAStringSet</a></code> object.</li>
<li>
<code>should_terminate</code>: A logical flag to signal to the
streaming framework that it should terminate. Note that this also means
the loader is in charge of the termination criterion.</li>
</ul>
<p>We decide to hard-code the loader to parse
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mn>10</mn><mn>4</mn></msup><annotation encoding="application/x-tex">10^4</annotation></semantics></math>
reads for each chunk.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input_fasta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"PETRI-seq_forward_reads.fa.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"posDemux"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">loader</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">state</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">n_reads_per_chunk</span> <span class="op">&lt;-</span> <span class="fl">1e4</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">state</span><span class="op">$</span><span class="va">output_table_initialized</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Starting to load sequences"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">chunk</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span></span>
<span>        filepath <span class="op">=</span> <span class="va">input_fasta</span>,</span>
<span>        format <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>        nrec <span class="op">=</span> <span class="va">n_reads_per_chunk</span>,</span>
<span>        skip <span class="op">=</span> <span class="va">state</span><span class="op">$</span><span class="va">total_reads</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="va">n_reads_in_chunk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">chunk</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">n_reads_in_chunk</span> <span class="op">==</span> <span class="fl">0L</span> <span class="op">&amp;&amp;</span> <span class="va">state</span><span class="op">$</span><span class="va">output_table_initialized</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># The case when the initial chunk is empty is given</span></span>
<span>        <span class="co"># special treatment since</span></span>
<span>        <span class="co"># we want to create the table regardless</span></span>
<span>        <span class="co"># No more reads to process, make the outer framework return</span></span>
<span>        <span class="va">final_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            state <span class="op">=</span> <span class="va">state</span>,</span>
<span>            sequences <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>            should_terminate <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>        <span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Done demultiplexing"</span><span class="op">)</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">final_res</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">state</span><span class="op">$</span><span class="va">total_reads</span> <span class="op">&lt;-</span> <span class="va">state</span><span class="op">$</span><span class="va">total_reads</span> <span class="op">+</span> <span class="va">n_reads_in_chunk</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        state <span class="op">=</span> <span class="va">state</span>,</span>
<span>        sequences <span class="op">=</span> <span class="va">chunk</span>,</span>
<span>        should_terminate <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-archiver">The archiver<a class="anchor" aria-label="anchor" href="#the-archiver"></a>
</h3>
<p>The archiver has the state obtained from the loader as its first
argument and the results from <code><a href="../reference/filter_demultiplex_res.html">filter_demultiplex_res()</a></code> of
the chunk as its second argument. Its return value is the state to
provide to the loader when it continues with its next chunk.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_umi_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"umi.fa"</span>, fileext <span class="op">=</span> <span class="st">".txt"</span><span class="op">)</span></span>
<span><span class="va">custom_output_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span></span>
<span>    pattern <span class="op">=</span> <span class="st">"barcode_table_custom"</span>,</span>
<span>    fileext <span class="op">=</span> <span class="st">".txt"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">archiver</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">state</span>, <span class="va">filtered_res</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">barcode_matrix</span> <span class="op">&lt;-</span> <span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">assigned_barcodes</span></span>
<span>    <span class="va">barcode_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">barcode_matrix</span><span class="op">)</span></span>
<span>    <span class="va">read_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">barcode_matrix</span><span class="op">)</span></span>
<span>    <span class="co"># If the table has no rows, we may risk getting a NULL value</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">read_names</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">read_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">barcode_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">barcode_matrix</span><span class="op">)</span></span>
<span>    <span class="va">read_name_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>read <span class="op">=</span> <span class="va">read_names</span><span class="op">)</span></span>
<span>    <span class="va">umis</span> <span class="op">&lt;-</span> <span class="va">filtered_res</span><span class="op">$</span><span class="va">demultiplex_res</span><span class="op">$</span><span class="va">payload</span><span class="op">$</span><span class="va">UMI</span></span>
<span></span>
<span>    <span class="va">chunk_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">read_name_table</span>, <span class="va">barcode_table</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">state</span><span class="op">$</span><span class="va">output_table_initialized</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">append</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span>        <span class="va">state</span><span class="op">$</span><span class="va">output_table_initialized</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">append</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">writeXStringSet</a></span><span class="op">(</span></span>
<span>        <span class="va">umis</span>,</span>
<span>        filepath <span class="op">=</span> <span class="va">output_umi_file</span>, append <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_tsv</a></span><span class="op">(</span></span>
<span>        x <span class="op">=</span> <span class="va">chunk_table</span>,</span>
<span>        file <span class="op">=</span> <span class="va">custom_output_table</span>,</span>
<span>        append <span class="op">=</span> <span class="va">append</span>,</span>
<span>        col_names <span class="op">=</span> <span class="op">!</span><span class="va">append</span>,</span>
<span>        eol <span class="op">=</span> <span class="st">"\n"</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="va">state</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">within</a></span><span class="op">(</span><span class="va">state</span>, <span class="op">{</span></span>
<span>        <span class="va">demultiplexed_reads</span> <span class="op">&lt;-</span> <span class="va">demultiplexed_reads</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">barcode_matrix</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>            <span class="st">"Processed {total_reads} reads,"</span>, <span class="st">" "</span>,</span>
<span>            <span class="st">"successfully demultiplexed {demultiplexed_reads} reads so far..."</span></span>
<span>        <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu">glue</span><span class="fu">::</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="va">state</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="putting-it-all-together">Putting it all together<a class="anchor" aria-label="anchor" href="#putting-it-all-together"></a>
</h3>
<p>With the initial state and the callbacks being defined, we are now
ready to put them to use:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_summary_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/streaming_demultiplex.html">streaming_demultiplex</a></span><span class="op">(</span></span>
<span>    state_init <span class="op">=</span> <span class="va">custom_state_init</span>,</span>
<span>    loader <span class="op">=</span> <span class="va">loader</span>,</span>
<span>    archiver <span class="op">=</span> <span class="va">archiver</span>,</span>
<span>    barcodes <span class="op">=</span> <span class="va">barcodes</span>,</span>
<span>    allowed_mismatches <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>    segments <span class="op">=</span> <span class="va">sequence_annotation</span>,</span>
<span>    segment_lengths <span class="op">=</span> <span class="va">segment_lengths</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting to load sequences</span></span>
<span><span class="co">#&gt; Processed 10000 reads, successfully demultiplexed 9117 reads so far...</span></span>
<span><span class="co">#&gt; Processed 20000 reads, successfully demultiplexed 18259 reads so far...</span></span>
<span><span class="co">#&gt; Processed 30000 reads, successfully demultiplexed 27362 reads so far...</span></span>
<span><span class="co">#&gt; Processed 40000 reads, successfully demultiplexed 36487 reads so far...</span></span>
<span><span class="co">#&gt; Processed 50000 reads, successfully demultiplexed 45629 reads so far...</span></span>
<span><span class="co">#&gt; Processed 56895 reads, successfully demultiplexed 51906 reads so far...</span></span>
<span><span class="co">#&gt; Done demultiplexing</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="verifying-results">Verifying results<a class="anchor" aria-label="anchor" href="#verifying-results"></a>
</h3>
<p>We finally load the files we wrote to in order to verify that they
contain what we desire. We load the barcode table:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">custom_output_table</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 51,906 × 4</span></span></span>
<span><span class="co">#&gt;    read   bc3    bc2    bc1   </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> seq_1  bc3_57 bc2_51 bc1_94</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> seq_2  bc3_72 bc2_94 bc1_95</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> seq_4  bc3_85 bc2_81 bc1_17</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> seq_5  bc3_49 bc2_64 bc1_77</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> seq_6  bc3_3  bc2_45 bc1_33</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> seq_7  bc3_17 bc2_57 bc1_71</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> seq_8  bc3_69 bc2_65 bc1_49</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> seq_9  bc3_38 bc2_87 bc1_83</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> seq_10 bc3_76 bc2_36 bc1_85</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> seq_11 bc3_15 bc2_84 bc1_16</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 51,896 more rows</span></span></span></code></pre></div>
<p>and the UMI sequences:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">output_umi_file</span>, format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 51906:</span></span>
<span><span class="co">#&gt;         width seq                                           names               </span></span>
<span><span class="co">#&gt;     [1]     7 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                       seq_1</span></span>
<span><span class="co">#&gt;     [2]     7 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_2</span></span>
<span><span class="co">#&gt;     [3]     7 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_4</span></span>
<span><span class="co">#&gt;     [4]     7 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                       seq_5</span></span>
<span><span class="co">#&gt;     [5]     7 <span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span>                                       seq_6</span></span>
<span><span class="co">#&gt;     ...   ... ...</span></span>
<span><span class="co">#&gt; [51902]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                       seq_56890</span></span>
<span><span class="co">#&gt; [51903]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span>                                       seq_56891</span></span>
<span><span class="co">#&gt; [51904]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_56892</span></span>
<span><span class="co">#&gt; [51905]     7 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span>                                       seq_56893</span></span>
<span><span class="co">#&gt; [51906]     7 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span>                                       seq_56895</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.23/posDemux" class="external-link">posDemux</a></em>
package was made possible thanks to:</p>
<ul>
<li>R</li>
<li><em><a href="https://bioconductor.org/packages/3.23/BiocStyle" class="external-link">BiocStyle</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=sessioninfo" class="external-link">sessioninfo</a></em></li>
<li><em><a href="https://CRAN.R-project.org/package=testthat" class="external-link">testthat</a></em></li>
</ul>
<p>This package was developed using <em><a href="https://bioconductor.org/packages/3.23/biocthis" class="external-link">biocthis</a></em>.</p>
<p><code>R</code> session information:</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R Under development (unstable) (2026-01-20 r89309)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2026-01-21</span></span>
<span><span class="co">#&gt;  pandoc   3.1.11 @ /opt/hostedtoolcache/pandoc/3.1.11/x64/ (via rmarkdown)</span></span>
<span><span class="co">#&gt;  quarto   NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package             </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version</span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  abind                  1.4-8   <span style="color: #555555;">2024-09-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  assertthat             0.2.1   <span style="color: #555555;">2019-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  backports              1.5.0   <span style="color: #555555;">2024-05-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bibtex                 0.5.1   <span style="color: #555555;">2023-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Biobase                2.71.0  <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocGenerics         * 0.57.0  <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocManager            1.30.27 <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  BiocParallel           1.45.0  <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocStyle            * 2.39.0  <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  Biostrings           * 2.79.4  <span style="color: #555555;">2026-01-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  bit                    4.6.0   <span style="color: #555555;">2025-03-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bit64                  4.6.0-1 <span style="color: #555555;">2025-01-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bitops                 1.0-9   <span style="color: #555555;">2024-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  blob                   1.3.0   <span style="color: #555555;">2026-01-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bookdown               0.46    <span style="color: #555555;">2025-12-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib                  0.9.0   <span style="color: #555555;">2025-01-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem                 1.1.0   <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  chunked              * 0.6.1   <span style="color: #555555;">2025-10-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cigarillo              1.1.0   <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  cli                    3.6.5   <span style="color: #555555;">2025-04-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  codetools              0.2-20  <span style="color: #555555;">2024-03-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.6.0)</span></span></span>
<span><span class="co">#&gt;  crayon                 1.5.3   <span style="color: #555555;">2024-06-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  DBI                  * 1.2.3   <span style="color: #555555;">2024-06-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dbplyr                 2.5.1   <span style="color: #555555;">2025-09-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  DelayedArray           0.37.0  <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  deldir                 2.0-4   <span style="color: #555555;">2024-02-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc                   1.4.3   <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest                 0.6.39  <span style="color: #555555;">2025-11-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dplyr                * 1.1.4   <span style="color: #555555;">2023-11-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate               1.0.5   <span style="color: #555555;">2025-08-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  farver                 2.1.2   <span style="color: #555555;">2024-05-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap                1.2.0   <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs                     1.6.6   <span style="color: #555555;">2025-04-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics             * 0.1.4   <span style="color: #555555;">2025-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  GenomicAlignments      1.47.0  <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  GenomicRanges          1.63.1  <span style="color: #555555;">2025-12-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  ggplot2                4.0.1   <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue                   1.8.0   <span style="color: #555555;">2024-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable                 0.3.6   <span style="color: #555555;">2024-10-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  hms                    1.1.4   <span style="color: #555555;">2025-10-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools              0.5.9   <span style="color: #555555;">2025-12-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets            1.6.4   <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httpuv                 1.6.16  <span style="color: #555555;">2025-04-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httr                   1.4.7   <span style="color: #555555;">2023-08-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  hwriter                1.3.2.1 <span style="color: #555555;">2022-04-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  interp                 1.1-6   <span style="color: #555555;">2024-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  IRanges              * 2.45.0  <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  jpeg                   0.1-11  <span style="color: #555555;">2025-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib              0.1.4   <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite               2.0.0   <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr                  1.51    <span style="color: #555555;">2025-12-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  LaF                    0.8.6   <span style="color: #555555;">2024-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  later                  1.4.5   <span style="color: #555555;">2026-01-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lattice                0.22-7  <span style="color: #555555;">2025-04-02</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.6.0)</span></span></span>
<span><span class="co">#&gt;  latticeExtra           0.6-31  <span style="color: #555555;">2025-09-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lifecycle              1.0.5   <span style="color: #555555;">2026-01-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lubridate              1.9.4   <span style="color: #555555;">2024-12-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr             * 2.0.4   <span style="color: #555555;">2025-09-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Matrix                 1.7-4   <span style="color: #555555;">2025-08-28</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.6.0)</span></span></span>
<span><span class="co">#&gt;  MatrixGenerics         1.23.0  <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  matrixStats            1.5.0   <span style="color: #555555;">2025-01-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  memoise                2.0.1   <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mime                   0.13    <span style="color: #555555;">2025-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  otel                   0.2.0   <span style="color: #555555;">2025-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pillar                 1.11.1  <span style="color: #555555;">2025-09-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig              2.0.3   <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown                2.2.0   <span style="color: #555555;">2025-11-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  plyr                   1.8.9   <span style="color: #555555;">2023-10-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  png                    0.1-8   <span style="color: #555555;">2022-11-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  posDemux             * 0.99.10 <span style="color: #555555;">2026-01-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  promises               1.5.0   <span style="color: #555555;">2025-11-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr                * 1.2.1   <span style="color: #555555;">2026-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pwalign                1.7.0   <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  R6                     2.6.1   <span style="color: #555555;">2025-02-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg                   1.5.0   <span style="color: #555555;">2025-09-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RColorBrewer           1.1-3   <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp                   1.1.1   <span style="color: #555555;">2026-01-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  readr                  2.1.6   <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RefManageR           * 1.4.0   <span style="color: #555555;">2022-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rlang                  1.1.7   <span style="color: #555555;">2026-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown              2.30    <span style="color: #555555;">2025-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rsamtools              2.27.0  <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  RSQLite                2.4.5   <span style="color: #555555;">2025-11-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  S4Arrays               1.11.1  <span style="color: #555555;">2025-11-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  S4Vectors            * 0.49.0  <span style="color: #555555;">2025-10-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  S7                     0.2.1   <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sass                   0.4.10  <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales                 1.4.0   <span style="color: #555555;">2025-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Seqinfo              * 1.1.0   <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  sessioninfo          * 1.2.3   <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shiny                  1.12.1  <span style="color: #555555;">2025-12-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ShortRead              1.69.2  <span style="color: #555555;">2025-11-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  SparseArray            1.11.10 <span style="color: #555555;">2025-12-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  stringi                1.8.7   <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr                1.6.0   <span style="color: #555555;">2025-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  SummarizedExperiment   1.41.0  <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  systemfonts            1.3.1   <span style="color: #555555;">2025-10-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping            1.0.4   <span style="color: #555555;">2025-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble                 3.3.1   <span style="color: #555555;">2026-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect             1.2.1   <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  timechange             0.3.0   <span style="color: #555555;">2024-01-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tzdb                   0.5.0   <span style="color: #555555;">2025-03-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  utf8                   1.2.6   <span style="color: #555555;">2025-06-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs                  0.7.0   <span style="color: #555555;">2026-01-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vroom                  1.6.7   <span style="color: #555555;">2025-11-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr                  3.0.2   <span style="color: #555555;">2024-10-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun                   0.56    <span style="color: #555555;">2026-01-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xml2                   1.5.2   <span style="color: #555555;">2026-01-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xtable                 1.8-4   <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  XVector              * 0.51.0  <span style="color: #555555;">2025-10-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  yaml                   2.3.12  <span style="color: #555555;">2025-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/devel/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/devel/lib/R/library</span></span></span>
<span><span class="co">#&gt;  <span style="color: #BBBBBB; background-color: #BB0000;">*</span> ── Packages attached to the search path.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.23/BiocStyle" class="external-link">BiocStyle</a></em>
with <em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em> and
<em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em>
running behind the scenes.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jakob Peder Pettersen, Centre for new antibacterial strategies (CANS).</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
